File: 1114_Print in Order/PrintinOrder.cs
using System.Threading;

public class Foo
{
    private EventWaitHandle waitFirst;
    private EventWaitHandle waitSecond;

    public Foo()
    {
        waitFirst = new AutoResetEvent(initialState: false);
        waitSecond = new AutoResetEvent(initialState: false);
    }

    public void First(Action printFirst)
    {
        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();

        waitFirst.Set();
    }

    public void Second(Action printSecond)
    {
        waitFirst.WaitOne();

        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();

        waitSecond.Set();
    }

    public void Third(Action printThird)
    {
        waitSecond.WaitOne();

        // printThird() outputs "third". Do not change or remove this line.
        printThird();
    }
}

File: 1114_Print in Order/PrintinOrder.java
class Foo {
    private Semaphore semaphore1, semaphore2;

    public Foo() {
        semaphore1 = new Semaphore(0);
        semaphore2 = new Semaphore(0);
    }

    public void first(Runnable printFirst) throws InterruptedException {
        printFirst.run();
        semaphore1.release();
    }

    public void second(Runnable printSecond) throws InterruptedException {
        semaphore1.acquire();
        printSecond.run();
        semaphore2.release();
    }

    public void third(Runnable printThird) throws InterruptedException {
        semaphore2.acquire(); 
        printThird.run();
        semaphore2.release();
    }
}